tests/CMakeLists.txt

```diff
@@ -310,8 +310,13 @@ function(pybind11_enable_warnings target_name)
         ${target_name}
         PRIVATE
           -Werror-all
-          # "Inlining inhibited by limit max-size", "Inlining inhibited by limit max-total-size"
-          -diag-disable 11074,11076)
+          # 2196: routine is both "inline" and "noinline" (no way to `pop` and not get the warning:
+          #       https://community.intel.com/t5/Intel-C-Compiler/Inline-and-no-inline-warning/td-p/1216764)
+          # 11074: Inlining inhibited by limit max-size
+          # 11074: Inlining inhibited by limit max-total-size
+          # 11076: To get full report use -qopt-report=4 -qopt-report-phase ipo
+          -diag-disable
+          2196,11074,11076)
```

Comment:
This should be added by the main CMakeLists to the correct target, not in tests. Someone using CMake (at least) should get the warnings added. ("This" being the inline/noinline one, not the max-size, as those can be handled by users)

Reply:
Could you please guide me how?

* "the main CMakeLists"  —  is that https://github.com/pybind/pybind11/blob/master/CMakeLists.txt?
* "correct target"  — I looked through top to bottom but didn't see how to find the "correct target". Somewhere around line 184? (I'm completely lost TBH.)

I have this: https://gist.github.com/rwgk/9c4305ccc492ad78654f842f58f36b09

It shows exactly which header files trigger `#2196`. Is that relevant here?

Reply:
Here, actually: https://github.com/pybind/pybind11/blob/master/tools/pybind11Common.cmake

I wish the names of the targets were a bit more general - windows helper, etc. are too specific to add this too. Maybe we need an intel_helper?

Reply:
> Here, actually: https://github.com/pybind/pybind11/blob/master/tools/pybind11Common.cmake
> 
> I wish the names of the targets were a bit more general - windows helper, etc. are too specific to add this too. Maybe we need an intel_helper?

Gulp ... will it be simpler like e.g.
```
set(clang_4plus
    "$<AND:$<CXX_COMPILER_ID:Clang>,$<NOT:$<VERSION_LESS:$<CXX_COMPILER_VERSION>,3.9>>>")
set(no_register "$<OR:${clang_4plus},$<CXX_COMPILER_ID:AppleClang>>")
```
?
If you could help out with this part that would be really awesome.

Reply:
Wait ... thinking about it ...

> This should be added by the main CMakeLists to the correct target, not in tests.

If that's what we want, we could just add this in common.h and be done with it:

```
#if defined(__INTEL_COMPILER)
// This pragma cannot be pop'ed:
// https://community.intel.com/t5/Intel-C-Compiler/Inline-and-no-inline-warning/td-p/1216764
#  pragma warning disable 2196  // warning #2196: routine is both "inline" and "noinline"
#endif
```

I'll make that revision asap.